binding.go

```diff
@@ -202,8 +205,23 @@ func Json(jsonStruct interface{}, ifacePtr ...interface{}) macaron.Handler {
 		ensureNotPointer(jsonStruct)
 		jsonStruct := reflect.New(reflect.TypeOf(jsonStruct))
 		if ctx.Req.Request.Body != nil {
-			defer ctx.Req.Request.Body.Close()
-			err := json.NewDecoder(ctx.Req.Request.Body).Decode(jsonStruct.Interface())
+			var err error
+			if ctx.Req.Method == "POST" || ctx.Req.Method == "PUT" || ctx.Req.Method == "PATCH" {
+				defer ctx.Req.Request.Body.Close()
```

Comment:
Put defer in `if-struct` could cause resource leak?

Reply:
We don't even need to close the body. See here: https://github.com/golang/go/blob/master/src/net/http/request.go#L178

Reply:
Thanks for the link, but I'm confused with the part "The Server will close the request body." which "server" it is?

I think we must make sure via source code line, not the description of the field (since this sounds critical to me).

Reply:
I have updated the pr.

Reply:
Hmm, I do not think I understand the logic here, what about GET/HEAD requests? We don't close body for requests other than POST/PUT/PATCH?

Reply:
Request Body is only expected and parsed for POST/PUT/PATCH.

Regarding Closing body, we don't need to do anything. Go handled that automatically https://stackoverflow.com/questions/42525499/where-to-put-defer-req-body-close . I closed the body for `POST/PUT/PATCH` case, but I can remove that line too and it will be ok.

You can see that the official ParseForm() method does not close body.
https://github.com/golang/go/blob/master/src/net/http/request.go#L1130:6

Reply:
Request Body is only expected and parsed for POST/PUT/PATCH.

Regarding Closing body, we don't need to do anything. Go handled that automatically https://stackoverflow.com/questions/42525499/where-to-put-defer-req-body-close . I closed the body for `POST/PUT/PATCH` case, but I can remove that line too and it will be ok.

You can see that the official ParseForm() method does not close body.
https://github.com/golang/go/blob/master/src/net/http/request.go#L1130:6

Reply:
OK. Let's delete this line!

Reply:
Done!